### DISCUSSION TITLE
Add a way to figure out if entity's `Damageable#damage` was cancelled

### DISCUSSION THREAD

**Discussion by imDaniX** [PaperMC/Paper#10905] [Ideas]
### Is your feature request related to a problem?

I need to react differently if entity with [`Damageable#damage`](https://jd.papermc.io/paper/1.21/org/bukkit/entity/Damageable.html#damage(double)) was damaged successfully or not. Currently, there's no way to figure it out, as damaging can be cancelled with `EntityDamageEvent#setCancelled(true)` with no feedback.

### Describe the solution you'd like.

Make `Damageable#damage` methods return boolean, with `false` on a cancelled event.

### Describe alternatives you've considered.

Calling `EntityDamageEvent` myself (which is unsupported, as events constructors are considered internal) and checking if it was cancelled.
I was thinking about checking entity's health before and after, but the final damage might result in 0, which if different from cancellation.

### Other

`Damageable#heal` might need a similar feature, since `EntityRegainHealthEvent` is cancellable too.

**Comment by electronicboy** [comment]
can't change the return type without breaking ABI

**Reply by imDaniX** [reply]
Then I think it's worth creating a new set of methods like `callDamage` (or, reflecting underneath methods, `damageEntity`) with the said or similar implementation. Then the original `damage` methods can be deprecated indefinitely, and their implementation to be changed to just calling the new methods.
There are plenty of plugins that do add custom weapons and stuff, and to properly consider for PvE/PvP protecting plugins like WorldGuard or CombatLogX, there's always a need for some tricky workarounds.

Otherwise, well, I think it should at least be noted in the javadocs that damaging can be cancelled.
